### PR TITLE
A tutorial on Build runner for galaxy

### PR DESCRIPTION
**Pull request by varunshankar**
Google Summer of Code 2016: Open genome informatics
Project Idea 13: Using Go-Docker to enhance Galaxy functionalities.

As part of the gsoc program, we have planned to give a detailed "step by step tutorial for building a runner for Galaxy". This tutorial will be highly useful for beginners to understand about the runner workflow, runner methods implementation and usage of galaxy.jobs package in respect with runner creation.

### CODE REVIEW

**Comment by dannon** [comment]
@varunshankar just wanted to chime in and say I love this idea, will give it a thorough review later!

**Comment by dannon** [comment]
I converted this to RST in https://github.com/varunshankar/galaxy/pull/1, which you'll want to use if you want this to show up in the documentation we build alongside Galaxy.

File: doc/source/dev/build_a_job_runner.rst [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+A walk through the steps of building a runner for Galaxy.
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In this tutorial, we would build the runner in a block by block fashion
+(like the building blocks), so we would divide the runner into
+components based on their function.
+
+Assuming you are familiar with setting up local install of galaxy and
+execution of galaxy.
```
Comment (nsoranzo) [inline doc/source/dev/build_a_job_runner.rst]: We assume you are familiar with setting up and managing a local installation of Galaxy.

File: doc/source/dev/build_a_job_runner.rst [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+In this tutorial, we would build the runner in a block by block fashion
+(like the building blocks), so we would divide the runner into
+components based on their function.
+
+Assuming you are familiar with setting up local install of galaxy and
+execution of galaxy. 
+
+To learn more about the basics, refer :
+https://wiki.galaxyproject.org/Admin/GetGalaxy#Install\_More\_Tools
```
Comment (nsoranzo) [inline doc/source/dev/build_a_job_runner.rst]: Just `https://wiki.galaxyproject.org/Admin/GetGalaxy`

File: doc/source/dev/build_a_job_runner.rst [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In this tutorial, we would build the runner in a block by block fashion
+(like the building blocks), so we would divide the runner into
+components based on their function.
+
+Assuming you are familiar with setting up local install of galaxy and
+execution of galaxy. 
+
+To learn more about the basics, refer :
```
Comment (nsoranzo) [inline doc/source/dev/build_a_job_runner.rst]: `please refer to:` (also few lines below)

File: doc/source/dev/build_a_job_runner.rst [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+To explore existing runners, refer:
+https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/runner
+
+What is required to make a runner for Galaxy?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`galaxy.jobs.runners.\_\_init\_\_.py <https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/runners/__init__.py>`__
+has the base runner implementation. To create a new runner, that base
+runner must be inherited and only certain functions need to be
```
Comment (nsoranzo) [inline doc/source/dev/build_a_job_runner.rst]: Substitute `functions` with `methods` (also in the next 2 paragraphs).

**Comment by varunshankar** [comment]
@nsoranzo .. Thanks for the comments. I have rectified the document accordingly.

File: doc/source/dev/build_a_job_runner.rst [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Input params: \* job\_state (Object of
+`galaxy.jobs.runners.AsynchronousJobState <https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/runners/__init__.py#L400>`__)
+
+Output params: AsynchronousJobState object
+
+| Without going into much detail, assume there is a queue to track the
+  status of every job. eg:
+| |image0|
```
Comment (nsoranzo) [inline doc/source/dev/build_a_job_runner.rst]: I think you are missing an image here.

File: doc/source/dev/build_a_job_runner.rst [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+Note: Iterating through the queue is already taken care by the
+framework.
+
+To inform galaxy about the status of the job: \* Get the job status from
+external runner using the job\_id \* Check if the job is
+queued/running/completed.. etc. A general structure is provided below.
+\* Call self.mark\_as\_finished(job\_state), if the job has been
+successfully executed. \* Call self.mark\_as\_failed(job\_state), if the
+job has failed during execution. \* To change state of a job, change
+job\_state.running and job\_state.job\_wrapper.change\_state()
```
Comment (nsoranzo) [inline doc/source/dev/build_a_job_runner.rst]: Should this be a list? This is not shown properly.

File: doc/source/dev/build_a_job_runner.rst [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+| Without going into much detail, assume there is a queue to track the
+  status of every job. eg:
+| |image0|
+
+The galaxy framework updates status of a job by iterating through the
+queue. During iteration, it calls check\_watched\_item with the job.
+Your responsibility will be to get status of execution of job from
+external runner and return the updated status of the job, and also to
+copy output files for completed jobs.
```
Comment (nsoranzo) [inline doc/source/dev/build_a_job_runner.rst]: Missing articles: "the status of a job", "During the iteration", "to get the status of execution of the job from the external runner", "copy the output files for the completed jobs".

**Comment by nsoranzo** [comment]
@varunshankar Thanks for the fixes, I've added a couple more comments, but it looks very good!

**Comment by varunshankar** [comment]
@nsoranzo Thanks for the comments. I have fixed the list, images and missing articles.

**Comment by dannon** [comment]
@varunshankar Thanks for writing this up, looks really helpful!

**Comment by varunshankar** [comment]
Thanks a lot for merging the document. Happy that this is my first contribution to galaxy project. Will love to contribute more

**Pull request merged**